### ISSUE TITLE
Segmentation fault on save screenshot

### ISSUE THREAD

**Issue by tsupplis** [dirkwhoffmann/virtualc64#458]
Hi Dirk,

I get a 100% reproducible segfault on save screenshot.
- 10.14.1
- VirtualC64 3.1

[virtualc64.log](https://github.com/dirkwhoffmann/virtualc64/files/2552574/virtualc64.log)
[crashdump.txt](https://github.com/dirkwhoffmann/virtualc64/files/2552575/crashdump.txt)

When under llvm:
```
ImageUtilities.185::screenshot()
Process 48619 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x11c235000)
    frame #0: 0x00000001000ee984 VirtualC64.bin`___lldb_unnamed_symbol5346$$VirtualC64.bin + 676
VirtualC64.bin`___lldb_unnamed_symbol5346$$VirtualC64.bin:
->  0x1000ee984 <+676>: movl   %edx, -0x8(%r12,%rcx,8)
    0x1000ee989 <+681>: leaq   (%rax,%rcx), %rsi
    0x1000ee98d <+685>: movl   %edx, -0x4(%r12,%rcx,8)
    0x1000ee992 <+690>: cmpq   $0x2, %rsi
Target 0: (VirtualC64.bin) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x11c235000)
  * frame #0: 0x00000001000ee984 VirtualC64.bin`___lldb_unnamed_symbol5346$$VirtualC64.bin + 676
    frame #1: 0x00000001000eea70 VirtualC64.bin`___lldb_unnamed_symbol5348$$VirtualC64.bin + 16
    frame #2: 0x00000001000ed5f8 VirtualC64.bin`___lldb_unnamed_symbol5344$$VirtualC64.bin + 664
    frame #3: 0x00000001000c559d VirtualC64.bin`___lldb_unnamed_symbol4647$$VirtualC64.bin + 93
    frame #4: 0x00000001000c7256 VirtualC64.bin`___lldb_unnamed_symbol4667$$VirtualC64.bin + 22
    frame #5: 0x00000001000d6ce7 VirtualC64.bin`___lldb_unnamed_symbol5018$$VirtualC64.bin + 39
    frame #6: 0x00007fff4e0b9546 AppKit`-[NSSavePanel _didEndSheet:returnCode:contextInfo:] + 96
    frame #7: 0x00007fff4e24c450 AppKit`NSWindowEndWindowModalSession + 338
    frame #8: 0x00007fff4e0bbbfd AppKit`-[NSSavePanel ok:] + 382
    frame #9: 0x00007fff4dc2a2cf AppKit`-[NSApplication(NSResponder) sendAction:to:from:] + 312
    frame #10: 0x00007fff4dc954b2 AppKit`-[NSControl sendAction:to:] + 86
    frame #11: 0x00007fff4dc953e4 AppKit`__26-[NSCell _sendActionFrom:]_block_invoke + 136
    frame #12: 0x00007fff4dc952e6 AppKit`-[NSCell _sendActionFrom:] + 178
    frame #13: 0x00007fff4dcc2355 AppKit`-[NSButtonCell _sendActionFrom:] + 97
    frame #14: 0x00007fff4dc93bf2 AppKit`-[NSCell trackMouse:inRect:ofView:untilMouseUp:] + 2376
    frame #15: 0x00007fff4dcc20a9 AppKit`-[NSButtonCell trackMouse:inRect:ofView:untilMouseUp:] + 698
    frame #16: 0x00007fff4dc9262f AppKit`-[NSControl mouseDown:] + 791
    frame #17: 0x00007fff4db6bd63 AppKit`-[NSWindow(NSEventRouting) _handleMouseDownEvent:isDelayedEvent:] + 5668
    frame #18: 0x00007fff4da9fd8b AppKit`-[NSWindow(NSEventRouting) _reallySendEvent:isDelayedEvent:] + 2319
    frame #19: 0x00007fff4da9f231 AppKit`-[NSWindow(NSEventRouting) sendEvent:] + 481
    frame #20: 0x00007fff4d93c4a8 AppKit`-[NSApplication(NSEvent) sendEvent:] + 336
    frame #21: 0x00007fff4d929cf1 AppKit`-[NSApplication run] + 755
    frame #22: 0x00007fff4d9193f7 AppKit`NSApplicationMain + 780
    frame #23: 0x000000010004ab29 VirtualC64.bin`___lldb_unnamed_symbol1873$$VirtualC64.bin + 9
    frame #24: 0x00007fff7d5be08d libdyld.dylib`start + 1
    frame #25: 0x00007fff7d5be08d libdyld.dylib`start + 1
```

In ImageUtilities.swift:

<img width="845" alt="image" src="https://user-images.githubusercontent.com/11035456/48060536-28393200-e1b4-11e8-943d-7c6d70bd1483.png">

[ImageUtilities.swift#L44](https://github.com/dirkwhoffmann/virtualc64/blob/c324f6e9ebb03f9b3961143bf6f6702b272ab22f/OSX/MacGUI/ImageUtilities.swift#L44)

Array boundary seems to be the problem, on line 44, 

```
        for i in 0 ... (w * h)-1 {
```
is a working fix.

**Comment by dirkwhoffmann** [comment]
Ups, I forgot to modify the code when I went back from 2048 x 4096 to 2048 x 2048 texture size 🙄.

**Comment by dirkwhoffmann** [comment]
Taking screenshots is working again. However, the user might be confused by the current behavior, because the texture is exported and not the frame buffer (the frame buffer can be copied nicely by Apples 'grab' tool). I.e., if all gimmicks are turned on, the texture will contain an upscaled and blurred image without CRT effects (the CRT effects are drawn after rasterization in the fragment shader, so they don't appear in any texture). Hence, the output looks like this: 

![untitled](https://user-images.githubusercontent.com/12561945/48076333-48351980-e1e5-11e8-8c8e-1aa8b491793f.png)

It's great for development though, because it allows us to inspect the internal texture pixel by pixel.

**Comment by dirkwhoffmann** [comment]
It should work in V3.1.1. A beta of this release is at 

http://www.dirkwhoffmann.de/virtualc64/VirtualC64_3.1.1_beta1.zip

**Comment by dirkwhoffmann** [comment]
Fixed in 3.1.1 which is now online.
